libs/agno/agno/os/app.py

```diff
@@ -587,48 +591,101 @@ def _auto_discover_databases(self) -> None:
         self.dbs = dbs
         self.knowledge_dbs = knowledge_dbs
 
-    def _register_db_with_validation(self, registered_dbs: Dict[str, Any], db: Union[BaseDb, AsyncBaseDb]) -> None:
+        # Initialize/scaffold all discovered databases
+        if self.auto_provision_dbs:
+            import asyncio
+            import concurrent.futures
+
+            try:
+                # If we're already in an event loop, run in a separate thread
+                asyncio.get_running_loop()
+
+                def run_in_new_loop():
+                    new_loop = asyncio.new_event_loop()
+                    asyncio.set_event_loop(new_loop)
+                    try:
+                        return new_loop.run_until_complete(self._initialize_databases())
+                    finally:
+                        new_loop.close()
+
+                with concurrent.futures.ThreadPoolExecutor(max_workers=1) as executor:
+                    future = executor.submit(run_in_new_loop)
+                    future.result()  # Wait for completion
+
+            except RuntimeError:
+                # No event loop running, use asyncio.run
+                asyncio.run(self._initialize_databases())
+
+    async def _initialize_databases(self) -> None:
+        """Initialize/scaffold all discovered databases by ensuring their tables are created."""
+        from agno.db.base import AsyncBaseDb, BaseDb
+        from agno.utils.log import log_debug, log_warning
+
+        all_dbs: List[Union[BaseDb, AsyncBaseDb]] = []
+
+        # Collect all database instances
+        for db_list in self.dbs.values():
+            all_dbs.extend(db_list)
+        for db_list in self.knowledge_dbs.values():
+            all_dbs.extend(db_list)
+
+        # Remove duplicates by object identity (same physical instance)
+        # Multiple instances can have the same db.id but different table configurations
+        seen_instances: set = set()
+        unique_dbs: List[Union[BaseDb, AsyncBaseDb]] = []
+        for db in all_dbs:
+            db_instance_id = id(db)
+            if db_instance_id not in seen_instances:
+                seen_instances.add(db_instance_id)
+                unique_dbs.append(db)
+
+        # Separate sync and async databases
+        sync_dbs = [(db.id, db) for db in unique_dbs if not isinstance(db, AsyncBaseDb)]
+        async_dbs = [(db.id, db) for db in unique_dbs if isinstance(db, AsyncBaseDb)]
```

Comment:
```suggestion
    async def _initialize_databases(self) -> None:
        """Initialize all discovered databases and create all Agno tables that don't exist yet."""
        from itertools import chain

        from agno.db.base import AsyncBaseDb, BaseDb
        from agno.utils.log import log_debug, log_warning

        # Collect all database instances and remove duplicates by identity
        unique_dbs = list(
            {
                id(db): db
                for db in chain(
                    chain.from_iterable(self.dbs.values()), chain.from_iterable(self.knowledge_dbs.values())
                )
            }.values()
        )

        # Separate sync and async databases
        sync_dbs: List[Tuple[str, BaseDb]] = []
        async_dbs: List[Tuple[str, AsyncBaseDb]] = []

        for db in unique_dbs:
            target = async_dbs if isinstance(db, AsyncBaseDb) else sync_dbs
            target.append((db.id, db))  # type: ignore
```


We can simplify and reduce iterations here